File: problems/First Bad Version/firstBadVersion.cs
/* The isBadVersion API is defined in the parent class VersionControl.
      bool IsBadVersion(int version); */

public class Solution : VersionControl {
    public int FirstBadVersion(int n) {
        var l = 1;
        var r = n;

        while (l <= r) {
            var mid = l + ((r - l) >> 1);

            if (IsBadVersion(mid)) {
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }

        return l;
    }
}

File: problems/First Bad Version/firstBadVersion.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        var left = 1;
        var right = n;

        while (left <= right) {
            var mid = left + ((right - left) >> 1);

            if (isBadVersion(mid)) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }
}
